gremlin-python/src/main/python/gremlin_python/driver/protocol.py

```diff
@@ -232,28 +233,27 @@ def data_received(self, response, results_dict):
 
         if response['ok']:
             message = self._message_serializer.deserialize_message(response['content'])
-            request_id = message['requestId']
-            result_set = results_dict[request_id] if request_id in results_dict else ResultSet(None, None)
             status_code = message['status']['code']
             aggregate_to = message['result']['meta'].get('aggregateTo', 'list')
             data = message['result']['data']
             result_set.aggregate_to = aggregate_to
 
             if status_code == 204:
                 result_set.stream.put_nowait([])
-                del results_dict[request_id]
                 return status_code
             elif status_code in [200, 206]:
                 result_set.stream.put_nowait(data)
-                if status_code == 200:
-                    result_set.status_attributes = message['status']['attributes']
-                    del results_dict[request_id]
+                # if status_code == 200:
+                #     result_set.status_attributes = message['status']['attributes']
                 return status_code
+            else:
+                log.error("\r\nReceived error message '%s'\r\n\r\nWith result set '%s'",
+                          str(message), str(result_set))
+                raise GremlinServerError({'code': status_code,
+                                          'message': message['status']['message'],
+                                          'attributes': {}})
         else:
-            # This message is going to be huge and kind of hard to read, but in the event of an error,
-            # it can provide invaluable info, so space it out appropriately.
-            log.error("\r\nReceived error message '%s'\r\n\r\nWith results dictionary '%s'",
-                      str(response['content']), str(results_dict))
+            log.error("\r\nReceived error message '%s'\r\n\r\nWith result set '%s'",
+                      str(response['content']), str(result_set))
             body = json.loads(response['content'])
```

Comment:
You could get an error that is serialized in GraphBinaryV4 back here. Should check the Content-Type instead of relying on it being JSON.

Reply:
Does the server serialize errors to GraphBinary? Even if I only have GraphBinaryV4 serializer enabled in server yaml, error responses are still sent as `application/json` (ok responses are in `application/vnd.graphbinary-v4.0`).

Reply:
It actually depends on the error and when the error occurs. Some are always serialized to JSON and other times it will serialize depending on the "Accept" header that was sent with the request. It really just depends on where the failure occurred in the server. For instance, if you sent an accept header that contains a MIME type the server can't handle, then obviously the server would have to send back JSON. But if you sent an error that occurred after processing has started, then it can return back serialized. In either case, the server should send back the correct content type for you to use.

Reply:
Ok, I dug a bit into this, so I think this might come from a combination of aiohttp defines "ok":
`Boolean representation of HTTP status code ([bool](https://docs.python.org/3/library/functions.html#bool)). True if status is less than 400; otherwise, False.` and how chunking isn't being handled right now.

Basically for now, I think due to not implementing chunking and allowing the library to handle it natively, we are only passing the first status we get from the server, which will be a 200, and that will make the final failing request fall into the `if response['ok']` case, and errors will be properly serialized. 

And the not `ok` cases look to be returned with `'Content-Type': 'application/vnd.graphbinary-v4.0'`, which the `else` case can handle.

Note that my request headers are:
```
CONTENT-TYPE: application/vnd.graphbinary-v4.0
ACCEPT: application/vnd.graphbinary-v4.0
```

For example (printing response headers and error output):
1. Tweaked serializer to send wrong number of bytes to server:
```
<CIMultiDictProxy('Content-Type': 'application/json', 'Content-Length': '81', 'Gremlin-RequestId': 'cf6793cd-316e-4f88-bdde-6d1d82b9a094')>
2024-06-21 12:38:15,918 [   ERROR] [    protocol.py:254 - data_received()] - 
Received error message 'b'{"message":"Unable to deserialize request using: GraphBinaryMessageSerializerV4"}''
```
2. Sent a math() step dividing by 0:
```
 <CIMultiDictProxy('Transfer-Encoding': 'chunked', 'Content-Type': 'application/vnd.graphbinary-v4.0', 'Gremlin-RequestId': '42fad392-12d0-4978-a085-7338897751d3', 'code': '500')>
Received error message '{'status': {'code': 500, 'message': 'Division by zero!', 'exception': 'ServerErrorException'}, 'result': {'meta': {}, 'data': []}}'
With result set '<gremlin_python.driver.resultset.ResultSet object at 0x10a42b6d0>'
```
3. Sent long request that caused server timeouts during processing:
```
<CIMultiDictProxy('Transfer-Encoding': 'chunked', 'Content-Type': 'application/vnd.graphbinary-v4.0', 'Gremlin-RequestId': '84d848f7-3747-4da6-83d1-93f90bc106f3', 'code': '500')>
2024-06-21 12:32:22,415 [   ERROR] [    protocol.py:248 - data_received()] - 
Received error message '{'status': {'code': 500, 'message': 'A timeout occurred during traversal evaluation of [RequestMessageV4{, fields={gremlinType=bytecode, g=g, bindings={}}, gremlin=[[], [V(), limit(10000)]]}]
```

So I think what I'll do is to add a `TODO` for checking on this when doing proper chunking to make sure errors are still properly serialized.